drivers/display/display_st7528i.c

```diff
@@ -0,0 +1,513 @@
+{
+	const struct st7528i_config *config = dev->config;
+	int ret;
+
+	if (config->height != ST7528_HEIGHT) {
+		LOG_ERR("Invalid display height: expected %d, got %d", ST7528_HEIGHT,
+			config->height);
+		return -ENODEV;
+	}
+
+	if (config->width != ST7528_WIDTH) {
+		LOG_ERR("Invalid display width: expected %d, got %d", ST7528_WIDTH, config->width);
+		return -ENODEV;
+	}
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus.bus->name);
+		return -ENODEV;
+	}
+
+	if (config->reset.port != NULL) {
+		if (!device_is_ready(config->reset.port)) {
+			LOG_ERR("Reset pin GPIO port not ready");
+			return -ENODEV;
+		}
+
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT_INACTIVE);
+
+		if (ret) {
+			LOG_ERR("Failed to configure reset pin");
+			return ret;
+		}
+	}
+
+	if (config->chip_select.port != NULL) {
+		if (!device_is_ready(config->chip_select.port)) {
+			LOG_ERR("Chip select pin GPIO port not ready");
+			return -ENODEV;
+		}
+
+		ret = gpio_pin_configure_dt(&config->chip_select, GPIO_OUTPUT_ACTIVE);
+
+		if (ret) {
+			LOG_ERR("Failed to configure chip select pin");
+			return ret;
+		}
+	}
+
+	ret = st7528i_reset(dev);
+	if (ret) {
+		LOG_ERR("reset failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline void st7528i_set_pixel(struct st7528i_data *data, const uint16_t col,
+				     const uint16_t row, const uint8_t level)
+{
+	__ASSERT(col < ST7528_WIDTH, "width out of bounds");
+	__ASSERT(row < ST7528_HEIGHT, "height out of bounds");
+
+	const int page = row / 8;
+	const int page_mask = 1 << (row % 8);
+	uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+
+	for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
+		if (level & (1 << level_bit)) {
+			ptr[level_bit] |= page_mask;
+		} else {
+			ptr[level_bit] &= ~page_mask;
+		}
+	}
+}
+
+static inline uint8_t st7528i_get_pixel(struct st7528i_data *data, const uint16_t col,
+					const uint16_t row)
+{
+	__ASSERT(col < ST7528_WIDTH, "width out of bounds");
+	__ASSERT(row < ST7528_HEIGHT, "height out of bounds");
+
+	const int page = row / 8;
+	const int page_mask = 1 << (row % 8);
+	uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+
+	uint8_t level = 0;
+
+	for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
+		if (ptr[level_bit] & page_mask) {
+			level |= (1 << level_bit);
+		}
+	}
+
+	return level;
+}
+
+static int st7528i_write_columns(const struct device *dev, const uint16_t page,
+				 const uint16_t start_col, const uint16_t num_cols)
+{
+	__ASSERT(page < ST7528_NUM_PAGES, "page out of bounds");
+	__ASSERT(start_col + num_cols <= ST7528_WIDTH, "columns out of bounds");
+
+	struct st7528i_data *data = dev->data;
+	int ret;
+
+	uint8_t cmd_buf[] = {
+		ST7528_CMD_SET_PAGE | page,
+		ST7528_CMD_SET_COL_MSN | (start_col >> 4),
+		ST7528_CMD_SET_COL_LSN | (start_col & 0x0f),
+	};
+
+	ret = st7528i_write_cmds(dev, cmd_buf, ARRAY_SIZE(cmd_buf));
```

Comment:
What is the reason to use `ARRAY_SIZE()` on `uint8_t cmd_buf[]`?
`ret = st7528i_write_cmds(dev, cmd_buf, sizeof(cmd_buf));`

Reply:
Why not? It's the same thing, except that it also ensures that `cmd_buf` didn't accidentally decay to a pointer.

Reply:
It is not the same, and also not from the code readability.